File: tests/python/greedy_algorithms/optimal_freelancing/test_optimal_freelancing.py
from problems.greedy_algorithms.optimal_freelancing.python.optimal_freelancing import \
    optimalFreelancing
from tests.python.utility import load_test_cases


def test_optimal_freelancing() -> None:
    test_cases = load_test_cases("greedy_algorithms/optimal_freelancing.json")

    for idx, test in enumerate(test_cases):
        jobs: list[dict] = test["jobs"]
        expected: int = test["expected"]

        result: int = optimalFreelancing(jobs)

        assert (
            result == expected
        ), f"Test case {idx} failed: expected {expected}, got {result}"

File: tests/cpp/greedy_algorithms/test_optimal_freelancing.cpp
#include "../../../problems/greedy_algorithms/optimal_freelancing/cpp/optimal_freelancing.h"
#include "../jsontestbase.h"

#include <nlohmann/json.hpp>
#include <unordered_map>
#include <vector>

using json = nlohmann::json;

class OptimalFreelancingTest : public JsonTestBase {
public:
  OptimalFreelancingTest() {
    json_file_path = std::string(TEST_CASES_DIR) +
                     "/greedy_algorithms/optimal_freelancing.json";
  }
};

std::vector<std::unordered_map<std::string, int>>
buildList(const json &jsonJobs) {
  std::vector<std::unordered_map<std::string, int>> jobs;
  for (const auto &job : jsonJobs) {
    std::unordered_map<std::string, int> jobMap;
    jobMap["deadline"] = job["deadline"];
    jobMap["payment"] = job["payment"];
    jobs.push_back(jobMap);
  }
  return jobs;
}

TEST_F(OptimalFreelancingTest, TestCases) {
  ASSERT_FALSE(test_cases.empty()) << "Test cases are empty or not loaded";

  for (const auto &test : test_cases) {
    auto jobs = buildList(test["jobs"]);
    int expected = test["expected"];

    auto result = optimalFreelancing(jobs);

    EXPECT_EQ(result, expected) << "Failed for test: " << test.dump(2);
  }
}
